### DISCUSSION TITLE
pods

### DISCUSSION THREAD

**Discussion by helmutkemper** [moby/moby#44681] [Show and tell]
I'm creating a framework for chaos test for microservices and I need to group the containers in a pod, k8s style, but I have no idea how to do this.

In the example below, it creates three `nats` instances and leaves them running for 60 minutes without doing anything else.

```go
        package localFolderProject
        
        import (
          "github.com/helmutkemper/chaos/factory"
          "testing"
          "time"
        )
        
        func TestDevOps_Linear(t *testing.T) {
          primordial := factory.NewPrimordial().
            NetworkCreate("chaos_network", "10.0.0.0/16", "10.0.0.1").
            Test(t)
        
          factory.NewContainerFromImage("nats:latest").
            Ports("tcp", 4222, 4222, 4223, 4224).
            Create("nats", 3).
            Start()
        
          if !primordial.Monitor(60 * time.Minute) {
            t.FailNow()
          }
        }
```

In the example below, I make a clone of the polar project:

```go
        package githubproject
        
        import (
          "github.com/helmutkemper/chaos/factory"
          "testing"
        )
        
        func TestLinear(t *testing.T) {
          // create a network
          factory.NewPrimordial().
            NetworkCreate("test_network", "10.0.0.0/16", "10.0.0.1").
            Test(t)
        
          // cloning polar project
          factory.NewContainerFromGit(
            "polar:latest",
            "https://github.com/polarstreams/polar.git",
          ).
            ReplaceBeforeBuild("Dockerfile", "./Dockerfile").
            EnvironmentVar(
              []string{
                //"BARCO_DEV_MODE=true",
                "POLAR_SHUTDOWN_DELAY_SECS=0",
                "POLAR_CONSUMER_ADD_DELAY_MS=5000",
                "POLAR_SEGMENT_FLUSH_INTERVAL_MS=500",
                "POLAR_BROKER_NAMES=delete_polar_0,delete_polar_1,delete_polar_2",
                "POLAR_ORDINAL=0",
              },
              []string{
                //"BARCO_DEV_MODE=true",
                "POLAR_SHUTDOWN_DELAY_SECS=0",
                "POLAR_CONSUMER_ADD_DELAY_MS=5000",
                "POLAR_SEGMENT_FLUSH_INTERVAL_MS=500",
                "POLAR_BROKER_NAMES=delete_polar_0,delete_polar_1,delete_polar_2",
                "POLAR_ORDINAL=1",
              },
              []string{
                //"BARCO_DEV_MODE=true",
                "POLAR_SHUTDOWN_DELAY_SECS=0",
                "POLAR_CONSUMER_ADD_DELAY_MS=5000",
                "POLAR_SEGMENT_FLUSH_INTERVAL_MS=500",
                "POLAR_BROKER_NAMES=delete_polar_0,delete_polar_1,delete_polar_2",
                "POLAR_ORDINAL=2",
              },
            ).
            Ports("tcp", 9250, 9250).
            Ports("tcp", 9251, 9251).
            Ports("tcp", 9252, 9252).
            EnableChaos(1, 1, 1, 0.0).
            Create("polar", 3).
            Start()
        }

```

Is there any example of how to group the containers in pods, k8s style?

**Comment by thaJeztah** [comment]
Effectively, a "pod" is a set of containers that have share namespaces; kubernetes uses a "pause" container to set up the initial namespaces, and then other containers join those namespaces to form a "pod".

A "somewhat" equivalent for that would be like this;

```bash
# create the pause container
docker run -dit --name pause_container --ipc=shareable busybox

# then create other container(s) to join it
docker run -d --name container_one \
    --ipc=container:pause_container \
    --network=container:pause_container \
    --pid=container:pause_container \
    --volumes-from=pause_container \
    nginx:alpine
```

After the above both containers share various namespaces, and effectively are a "pod";

```bash
docker exec -it container_one ps aux
PID   USER     TIME  COMMAND
    1 root      0:00 sh
    7 root      0:00 nginx: master process nginx -g daemon off;
   36 nginx     0:00 nginx: worker process
   37 nginx     0:00 nginx: worker process
   38 nginx     0:00 nginx: worker process
   39 nginx     0:00 nginx: worker process
   40 nginx     0:00 nginx: worker process
   54 root      0:00 ps aux

docker exec -it pause_container ps aux
PID   USER     TIME  COMMAND
    1 root      0:00 sh
    7 root      0:00 nginx: master process nginx -g daemon off;
   36 101       0:00 nginx: worker process
   37 101       0:00 nginx: worker process
   38 101       0:00 nginx: worker process
   39 101       0:00 nginx: worker process
   40 101       0:00 nginx: worker process
   48 root      0:00 ps aux
```

And (for example) you can access the nginx on localhost from either container;

```bash
docker exec -it container_one sh -c 'wget -q -O- localhost | grep Welcome'
<title>Welcome to nginx!</title>
<h1>Welcome to nginx!</h1>
```

However, the docker daemon itself does not have a "pod" API; if you must have the "exact" equivalent of what Kubernetes does, you could have a look at https://github.com/Mirantis/cri-dockerd, which implements the Kubernetes CRI API for the Docker Daemon.

**Comment by helmutkemper** [comment]
I think I'm a little confused about my real need

Is this a stack?
<img width="528" alt="Captura de Tela 2022-12-22 às 11 21 30" src="https://user-images.githubusercontent.com/1098757/209156286-498786b3-0e00-4f27-abf9-28fc7932462c.png">

How do I do this with golang code?

**Reply by thaJeztah** [reply]
That looks like a regular compose project that's running; compose uses labels on the containers it creates to correlate them with a project, and it creates a custom network for the project, but otherwise they're just 3 regular containers.


You can find the metadata / labels it sets by inspecting those containers (`docker container inspect mongo3`); without the labels, it would be "roughly" the equivalent of;


```bash
docker network create --name myproject

docker run -d --name mongo1 --network=myproject --network-alias mongo -p 30001:3000 mongo:5
docker run -d --name mongo2 --network=myproject --network-alias mongo -p 30002:3000 mongo:5
docker run -d --name mongo3 --network=myproject --network-alias mongo -p 30003:3000 mongo:5
```

But to get a full equivalent, you may end up reinventing compose (https://github.com/docker/compose/), so perhaps shelling out to run compose could be an option? I see you're using some code to replace values of environment variables (e.g. `POLAR_SHUTDOWN_DELAY_SECS`); you can use build-args for those, and compose can use environment variables from the current environment, so in that case, compose could fill in those values.

I see you're using

**Comment by helmutkemper** [comment]
The idea of the project is not to rewrite docker compose. The idea of the project is to make a chaos testing tool aimed at developing microservices.

This is the third version of the tool, as the previous version allowed many configurations and was complicated to use.

The basic idea is to allow writing a test code to run on the developer's machine and to allow a small integration test, where the containers behave randomly, stopping, pausing, changing the IP address, and with this, allowing the developer to know how your microservice behaves.

For example, if the text "panic:" appears in the container's standard output, the container's standard output is saved on the host computer, in a folder named "bug", thus allowing a later analysis of the problem.
It also automatically saves inspection shots of containers in the form of CSV files, making it possible to measure memory and cpu consumption during testing.

My current problem, is that polar project, an alternative to kafka written in golang, was made for kubernetes and I have no idea how to group all instances of polar project into a single block and then connect to it by just passing a single block address.

For example, when I connect to the address polar://polar-service, it should connect to the three instances of polar, that is, polar://polar-container-0;polar-container-1;polar-container-2

I'm lost on how to do this without using kubernetes

[thaJeztah](https://github.com/thaJeztah), thank you very much for your help
